### PR TITLE
Reduce allocations with internal smart buffering

### PR DESCRIPTION
**Pull request by willardf**
Been thinking about them allocations lately...

Sucks that we make a bunch of new `byte[]` every time we do a send. But it's also annoying to have to deal with threaded reliable resends making buffer lifetimes racy.

So I reinvented Smart Pointers and it seems like it works pretty well.

Edit: To avoid confusion:
AddUsage is to be used right before a SmartBuffer needs to be preserved for some reason. This only happens in a few places: As we enqueue data for later, as we store data for resending, and as we call BeginSendTo.
Nearly all usages of SmartBuffer don't need AddUsage or Recycle. A simple `using` statement wraps that up so we can never fail to recycle.

### CODE REVIEW

File: Hazel/Dtls/DtlsConnectionListener.cs [hunk 1]
```diff
@@ -1275,13 +1285,13 @@ private void SendHelloVerifyRequest(IPEndPoint peerAddress, ulong recordSequence
...
 
-            base.QueueRawData(packet, peerAddress);
+            base.QueueRawData(buffer, peerAddress);
         }
 
         /// <summary>
         /// Handle a requrest to send a datagram to the network
         /// </summary>
-        protected override void QueueRawData(ByteSpan span, IPEndPoint remoteEndPoint)
+        protected override void QueueRawData(SmartBuffer span, IPEndPoint remoteEndPoint)
```
Comment (willardf) [inline Hazel/Dtls/DtlsConnectionListener.cs:1294]: This is the only method that scares me. It's really hard to trace the access pattern because DtlsConnectionListener very obvious uses base.QueueRawData to _not_ call this version, and that's good, but if something elsewhere doesn't call AddUsage once and accidentally gets this version, then the counter will go negative and bad things will happen.

Comment (willardf) [reply Hazel/Dtls/DtlsConnectionListener.cs:1294]: K, yeah, I'm pretty sure this was just a memory leak. So I fixed that and added some asserts to tests so after sending stuff, I'm very sure we sit at 0 buffers in use.

File: Hazel/Dtls/DtlsConnectionListener.cs [hunk 2]
```diff
@@ -1331,8 +1341,10 @@ protected override void QueueRawData(ByteSpan span, IPEndPoint remoteEndPoint)
                         ref outgoingRecord
                     );
 
-                    base.QueueRawData(packet, remoteEndPoint);
+                    queuedSpan.Recycle();
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsConnectionListener.cs:1344]: I don't think I see the associated AddUsage() call for this one - I'd assume it should be the 2nd line of the for block?

File: Hazel/Dtls/DtlsConnectionListener.cs [hunk 3]
```diff
@@ -1295,10 +1306,7 @@ protected override void QueueRawData(ByteSpan span, IPEndPoint remoteEndPoint)
                 // If we're negotiating a new epoch, queue data
                 if (peer.Epoch == 0 || peer.NextEpoch.State != HandshakeState.ExpectingHello)
                 {
-                    ByteSpan copyOfSpan = new byte[span.Length];
-                    span.CopyTo(copyOfSpan);
-
-                    peer.QueuedApplicationDataMessage.Add(copyOfSpan);
+                    peer.QueuedApplicationDataMessage.Add(span);
```
[the remark this answers is not in the document]
Comment (Mukikaizoku) [reply Hazel/Dtls/DtlsConnectionListener.cs:1309]: Ah - I see why it doesn't have one - because if we Recycle after queuing it, no other context has claimed a Usage to it and the object would be recycled before it's processed elsewhere

However, if that's the case, this content never releases its Usage call so I think that would prevent Recycling and cause a leak?

File: Hazel/Dtls/DtlsConnectionListener.cs [hunk 4]
```diff
@@ -1331,8 +1341,10 @@ protected override void QueueRawData(ByteSpan span, IPEndPoint remoteEndPoint)
                         ref outgoingRecord
                     );
 
-                    base.QueueRawData(packet, remoteEndPoint);
+                    queuedSpan.Recycle();
+                    base.QueueRawData(buffer, remoteEndPoint);
                 }
+
                 peer.QueuedApplicationDataMessage.Clear();
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsConnectionListener.cs:1348]: For the Clear(), I think we'd need to recycle all these objects before Clearing?

File: Hazel/Dtls/DtlsConnectionListener.cs (hunk 2 above)
Comment (Mukikaizoku) [reply Hazel/Dtls/DtlsConnectionListener.cs:1344]: Ahhh I think I might see how this works - this queue is only accessed by this same method so you can ensure the same number of AddUsage()/Recycle() pairs

File: Hazel/Dtls/DtlsUnityConnection.cs [hunk 5]
```diff
@@ -349,25 +352,31 @@ protected override void HandleSend(byte[] data, byte sendOption, Action ackCallb
         }
 
         /// <inheritdoc />
-        protected override void WriteBytesToConnection(byte[] bytes, int length)
+        protected override void WriteBytesToConnection(SmartBuffer bytes, int length)
         {
-            ByteSpan wireData = this.WriteBytesToConnectionInternal(bytes, length);
+            SmartBuffer wireData = this.WriteBytesToConnectionInternal(bytes, length);
+            bytes.Recycle();
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsUnityConnection.cs:358]: This assumes that something called AddUsage() on the SmartBuffer before this method was called right?

Comment (Mukikaizoku) [reply Hazel/Dtls/DtlsUnityConnection.cs:358]: I thiiiink in the call sites below, AddUsage() might not be called

File: Hazel/FewerThreads/ThreadLimitedUdpConnectionListener.cs [hunk 6]
```diff
@@ -262,6 +262,10 @@ private void SendLoop()
                     this.Logger.WriteError("Error in loop while sending: " + e.Message);
                     Thread.Sleep(1);
                 }
+                finally
+                {
+                    msg.Span.Recycle();
```
Comment (Mukikaizoku) [inline Hazel/FewerThreads/ThreadLimitedUdpConnectionListener.cs:267]: Here too - seems like it's vital that AddUsage() is called somewhere else

**Review by Mukikaizoku — COMMENTED** [review]
Alright, took a look and I think the goal to reduce allocations is met, but I'm not sure if I fully grasp the intended usage of SmartBuffer/had concerns

I might be just misunderstanding the SmartBuffer, but I do have concerns that this code could be introducing a few memory leaks. To avoid the leak, it seems the SmartBuffer needs to reach exactly 0 for usage count and it seems great care needs to be taken to be sure the call paths always call AddUsage() at least once. The tricky part I find is that the AddUsage()/Recycle() pair might live in different locations

If my understanding is correct, my recommendation would be to remove the usagecount and AddUsage() method. It seemed to me the use cases here are relatively isolated and don't need to support multiple contexts holding references simultaneously?

File: Hazel/Udp/UnityUdpClientConnection.cs [hunk 7]
```diff
@@ -85,19 +85,20 @@ protected override void WriteBytesToConnection(byte[] bytes, int length)
...
+                    (byte[])bytes,
                     0,
                     length,
                     SocketFlags.None,
                     EndPoint,
                     HandleSendTo,
-                    null);
+                    bytes);
             }
             catch (NullReferenceException) { }
```
Comment (blurdot) [inline Hazel/Udp/UnityUdpClientConnection.cs:103]: If this throws and the AsyncCallback doesn't fire, is the Connection always just disposed? If not we may need to make sure this SmartBuffer gets recycled in the catch blocks...

**Review by blurdot — COMMENTED** [review]
I like this! Had some questions, mebbe probably they could be dumb, but there were a few places where it was hard to follow the AddUsage/Recycle calls so I just went ahead and dropped questions where I had them. :)

File: Hazel/FewerThreads/ThreadLimitedUdpConnectionListener.cs [hunk 8]
```diff
@@ -334,13 +340,14 @@ protected virtual void ReadCallback(MessageReader message, IPEndPoint remoteEndP
...
+        internal void SendDataRaw(SmartBuffer response, IPEndPoint remoteEndPoint)
         {
             QueueRawData(response, remoteEndPoint);
         }
 
-        protected virtual void QueueRawData(ByteSpan span, IPEndPoint remoteEndPoint)
+        protected virtual void QueueRawData(SmartBuffer span, IPEndPoint remoteEndPoint)
         {
+            span.AddUsage();
             this.sendQueue.TryAdd(new SendMessageInfo() { Span = span, Recipient = remoteEndPoint });
```
[the remark this answers is not in the document]
Comment (willardf) [reply Hazel/FewerThreads/ThreadLimitedUdpConnectionListener.cs:351]: Naw, send queue joins, so the sender thread should best effort through the queue. In the off chance that fails, the intent is that if Listener is disposed, it won't be used anymore, so GC can collect ALL the buffers. Best part of having an instance buffer pool instead of a static one.

**Comment by willardf** [comment]
> If my understanding is correct, my recommendation would be to remove the usagecount and AddUsage() method. It seemed to me the use cases here are relatively isolated and don't need to support multiple contexts holding references simultaneously?

MAYBE, but probably you are missing a scenario. The issue that causes all of this is:

```
UnreliableSend -> BufferCopy -> BeginSendTo -> EndSendTo -> Recycle
ReliableSend -> BufferCopy -> AttachReliableId -> RetrySending -> BeginSendTo -> EndSendTo -> NO Recycle
                                                                              -> BeginSendTo -> EndSendTo -> NO Recycle
                                                                                                                                 -> ReceiveAck -> Recycle
```

In particular, we don't have a good way to differentiate a recycling send and a non-recycling send. Listeners have trouble with this because we queue data to another thread. But it's simple enough, we could pass the send option through the stack. Could even be part of the SmartBuffer... /thinking
But the scenario you might have missed is that we could be resending while we receive an ack. If we return the buffer we might send torn data. Buffers are shared across users, so it could even be a security issue.

That SAID! I think we can use this to simplify AddUsage. We don't need to use it _all the time_, just in PingPacket. Everywhere else should be allowed to get away with one Recycle and no AddUsage.

I will defend usageCount tho. Pooling _has_ to be precise. That's really the problem with pooling. If you recycle twice, that's never not an issue. The only bad part is that I should have made it explode if we reach negative.

File: Hazel/Udp/UnityUdpClientConnection.cs (hunk 7 above)
Comment (willardf) [reply Hazel/Udp/UnityUdpClientConnection.cs:103]: Most likely yeah, but good to throw the recycles in there anyway.

File: Hazel/Dtls/DtlsConnectionListener.cs (hunk 4 above)
Comment (willardf) [reply Hazel/Dtls/DtlsConnectionListener.cs:1351]: Refactored this to be a real queue so we feel more confident that its contents are always recycled.

File: Hazel/Dtls/DtlsConnectionListener.cs (hunk 2 above)
Comment (willardf) [reply Hazel/Dtls/DtlsConnectionListener.cs]: Refactored this away by with the using pattern

File: Hazel/Dtls/DtlsConnectionListener.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (willardf) [reply Hazel/Dtls/DtlsConnectionListener.cs:1313]: Yeah, broadly, AddUsing should always be called right before submitting a buffer to something that will store it for any amount of time (typically a queue or async method). The trick with how this was written is that once you dequeue it, you typically want to use it, THEN recycle it. So the usage becomes less obvious. Recycle appears in weird places unrelated to the rule of storing.

So I caved an implemented IDispose on SmartBuffer so we can use using declarations. It's the exact behaviour we need and only breaks down in the two spots that need special attention: PingPacket and SendTo.

File: Hazel/Udp/UnityUdpClientConnection.cs [hunk 9]
```diff
@@ -151,6 +157,10 @@ private void HandleSendTo(IAsyncResult result)
             {
                 DisconnectInternal(HazelInternalErrors.SocketExceptionSend, "Could not send data as a SocketException occurred: " + ex.Message);
             }
+            finally
+            {
+                ((SmartBuffer)result.AsyncState).Recycle();
```
[the remark this answers is not in the document]
Comment (willardf) [reply Hazel/Udp/UnityUdpClientConnection.cs:167]: No, the matching AddUsage is before we send. Clarified in my comment further down and I'll edit my description, but AddUsage is for preserving a buffer over an unknown period of time. (ie putting it in a queue or async method).

File: Hazel/SmartBuffer.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            return b.buffer;
+        }
+
+        public void AddUsage()
+        {
+            Interlocked.Increment(ref this.usageCount);
+        }
+
+        public void Recycle()
```
[the remark this answers is not in the document]
Comment (willardf) [reply Hazel/SmartBuffer.cs:62]: I'd like to not reimplement ObjectPool+IRecyclable, so Recycle has to stay. I could change AddUsage to something like Pin or something, but yeah, I'll probably not be able to budge here.

**Comment by willardf** [comment]
> Yeah, this all looks good to me. I would want to put it through its paces with some stress tests and probably a canary deployment to West-1?
> 
> @willardf, My inclination is that this should go into both the AU client and the AU server simultaneously, but mebbe we could we roll it out in the server first and test?

Yeah, because the changes are all internal, it should be totally safe to do a server deployment first. Worth noting that we've been running this version of Hazel (non-DTLS) in REDACTED since we did the optimization pass. So while not the paces, it's at least been smoke tested.
